aliases support

mg979 opened the pull request:
Works with toggle project and tested with all options

mg979 committed:
aliases support

mg979 commented on favorites.py:
```diff
@@ -179,10 +198,11 @@ def load_favorite_files(cls, obj, force=False, clean=False, win_id=None):
         # Is project enabled
         FavProjects.project_adjust(obj, win_id, force)
 
+        cls.check_aliases(obj.file_name)
```
I added the check here so that it's run indifferently for global and project json files

mg979 commented on favorites.py:
```diff
@@ -103,6 +103,25 @@ class FavFileMgr(object):
+            with open(favs, "w") as file:
+                json.dump(data, file, sort_keys=True, indent=4, separators=(',', ': '))
+        f.close()
+
```
function to update json files

mg979 commented on favorites.py:
```diff
@@ -179,10 +198,11 @@ def load_favorite_files(cls, obj, force=False, clean=False, win_id=None):
             if force:
                 # Create file list if it doesn't exist
-                if cls.create_favorite_list(obj, {"version": 1, "files": [], "groups": {}}, force=True):
+                if cls.create_favorite_list(obj, {"version": 1.1, "files": [], "groups": {}}, force=True):
```
In the function the version number is checked

facelessuser replied:
Lets just call it version `2`.  It is unlikely for formats that I am dong a `0.0.0` scheme.  I may use `0.0`, but this is clearly a major change, so let's call it `2`.

mg979 commented on favorites.py:
```diff
@@ -272,29 +294,33 @@ def exists(self, s, group=False, group_name=None):
-                return True if s in set(self.obj.files["groups"][group_name]) else False
+                i = [i for i, f in enumerate(self.obj.files["groups"][group_name]) if f["file"] == s]
+                return i[0] if i else False
 
```
changed to return the index because it's not a single string anymore, this is only used by remove because it needs the index

facelessuser replied:
I don't really care for this approach.  If it returns the index, then it should be called `get_index` or something like that. Also, if it is returning the index, it should return `None` if there is no index.  Also make sure that anyone who calls it is checking for `index is not None` and not for it's "truthiness" as an index of `0` as `0 == False` and is not `0 is False`.

If we are changing the the definition of what a function is doing, the name probably should change as well.

mg979 replied:
Ok for everything else(renaming file to filename, using version 2, sanitize json).

This is a bit more problematic, for now I separated the function for group and file check, since exists() is also an import from os and they can be mistaken for each other, I renamed them in group_exists() (returns True or False) and file_index(returns index or None).

> Also make sure that anyone who calls it is checking for index is not None and not for its "truthiness" as an index of 0 as 0 == False and is not 0 is False.

I was already checking `if index is not False` where needed(line 310), now changing to `if index is not None`. But while the returned index is an integer and can be zero, it is returned if `i` is not null, and `i` itself is a list of indexes(line 302), so it wasn't causing problems even if it only contained [0]. `i` is only null if the list is empty, that is if the file isn't there. Also checked in the interpreter that [0] doesn't return False. Anyway I'm separating functions to avoid misinterpretations, if you agree.

facelessuser replied:
Okay, I took a closer look.  There should be a function if `group_exists` and `get_index` or something like that. I don't like the function conditionally returning different things. This will be confusing for people coming behind to make changes.

They should return either a boolean, or an index, but not both.  If it needs to be two separate functions, than that is fine.  I didn't realize that it checks if group *or* file exists (I was just looking at the diff).  These should now be two separate functions.

Also, I much prefer if it is returning a index to either return something like `-1` or `None` instead of `False`.

mg979 replied:
I'm changing the function to:

```
    def exists(self, s, group=False, group_name=None):
        """Check if group or file exists."""

        if group:
            # See if group exists
            return True if s in self.obj.files["groups"] else False

        else:

            def is_present_in(key):
                for index, filename in enumerate(key):
                    if key['file'] == s:
                        return index
                return None

            # See if file in global or group list exists, and return its index
            if group_name is None:
                index = is_present_in(self.obj.files["files"])
            else:
                index = is_present_in(self.obj.files["groups"][group_name])

            return index if index is not None else None
```

So, no different names, just a bit clearer. Ok?

facelessuser replied:
But the function is  doing two different things now.  That's the problem.  In one case it is checking if it exists, in the second case it is actually returning/getting an index.

It's fine to have exists still work the same, you could even leverage the new `get_index` function.

```py
    def exists(self, s, group=False, group_name=None):
        """Check if group or file exists."""

        if group:
            # See if group exists
            return True if s in self.obj.files["groups"] else False

        else:

            return self.get_index(s, group, group_name) is not None
```

But I don't like it doing two different things.  You should call `self.get_index` when you want an index, and `self.exists` if you just want to know if it exists.  This will make it easier to for new contributors to understand in the future.

mg979 replied:
I'm changing the function to:

```
    def group_exists(self, s):
        """Check if group exists."""

        return True if s in self.obj.files["groups"] else False

    def file_index(self, s, group_name=None):
        """Check if file exists, and return its index in this case."""

        def is_present_in(key):
            for index, filename in enumerate(key):
                if filename['file'] == s:
                    return index
            return None

        if group_name is None:
            index = is_present_in(self.obj.files["files"])
        else:
            index = is_present_in(self.obj.files["groups"][group_name])

        return index if index is not None else None
```

facelessuser replied:
Technically you can just return index as index is already either an index or None.

Also the parameter in `is_present_in` is technically not a key, but the value of the dictionary.

I know I did it this way `return True if s in self.obj.files["groups"] else False` but we can move to just `return s in self.obj.files["groups"]`.  The result is already a boolean. The whole if else isn't really needed. I was still learning Python and did unnecessary things here and there when I wrote this.

But yeah, generally, this looks good 🙂 .

facelessuser replied:
The linter might gripe that `is_present_in` is missing a docstring.  It's not really a public function, but the linter will gripe anyways.  Just an FYI.

mg979 replied:
> Technically you can just return index as index is already either an index or None

I don't think so because if the index is zero it must be returned anyway in this case. If is_present_in() returns index zero, it must be returned.

> Also the parameter in is_present_in is technically not a key, but the value of the dictionary

Ok then value

> we can move to just return s in self.obj.files["groups"]

Ok. I'm still learning stuff myself, I'm not that advanced.

> The linter might gripe that is_present_in is missing a docstring

Will add.

facelessuser replied:
> I don't think so 

Maybe I wasn't clear what I was referring to:

```
return index if index is not None else None
```

If index is not None you are returning index.  If index is None, you are returning None.

So why not `return index`?  This is what I'm pointing out.  This line is redundant. Just return index.

mg979 replied:
Yes I understood now, you're right. But maybe it's better like this, I can avoid checking `if index is not None` in the calling function, because zero would be also a valid value.

facelessuser replied:
I don't understand. Index == 0 is valid, I get that.

But:

```py
>>> index = 0
>>> index if index is not None
0
>>> index
0
>>> index = None
>>> index if index is not None
None
>>> index
None
```

You still get 0 - x, and `None` if it's `None`.  And you still have to check  check if index is `None` in the calling function. You aren't gaining anything by keeping it in `file_index`. If you don't want it to return `None`, the only option is to raise an exception and catch the exception or return an invalid index like `-1`.  But you still have to check if value is `>=0`.

mg979 replied:
Yes sorry. Totally right, the check is still needed. Sorry again.

facelessuser replied:
No worries, like I said, there are probably still places in the code where I do this exact thing because I was learning 🙂 .

mg979 replied:
I could push these changes, and I also added a check to remember project mode across session. I can include it now or leave it for later, what should I do?

facelessuser replied:
Let's handle persistent per project stuff separately.

mg979 commented on favorites.py:
```diff
@@ -272,29 +294,33 @@ def exists(self, s, group=False, group_name=None):
+            return [[path['alias'], path['file']] for path in self.obj.files["groups"][group_name]]
         else:
-            return [[basename(path), path] for path in self.obj.files["files"]]
+            return [[path['alias'], path['file']] for path in self.obj.files["files"]]
```
updated for quick panel

mg979 commented:
I tested all options and they work, and the old jsons should be updated on the fly, global and project ones. I kept changes to the barest minimum possible, at least for me.

I also have another uncommitted commit ready that reworks some routines in AddFavoriteFileCommand, but I'll wait for your impressions on this one.

facelessuser commented:
In general, I like this a lot more 🙂 .  This is much cleaner.  Obviously we'll need to update to allow access of the alias before we merge, but I it sounds like you have plans to do that.  So looking good so far.

facelessuser commented:
So how does this get displayed now in the panel?

```
file or alias
full path
```

mg979 committed:
split exists function for file/group check

mg979 commented on favorite_files.py:
```diff
@@ -138,7 +142,7 @@ def create_group(self, value):
             repeat = True
-        elif Favs.exists(value, group=True):
+        elif Favs.group_exists(value):
             # Do not allow duplicates
```
functions renaming, groups_exists() doesn't need the argument 'group' anymore

mg979 commented on favorites.py:
```diff
@@ -106,20 +106,20 @@ class FavFileMgr(object):
+                data['groups'][group] = [{"file": filename, "alias": basename(filename)}
+                                         for filename in data['groups'][group]]
+            data["version"] = 2
             with open(favs, "w") as file:
```
filename instead of file, version 2 instead of 1.1

mg979 commented on favorites.py:
```diff
@@ -287,31 +287,38 @@ def set(self, s, group_name=None):
+        """Check if file exists, and return its index in this case."""
+
+        def is_present_in(files_list):
+            """Returns the index if the file is present"""
```
I forgot a period here and the Travis build failed.

facelessuser commented on favorites.py:
```diff
@@ -260,41 +280,54 @@ def add_group(self, s):
+        """Check if file exists, and return its index in this case."""
+
+        def is_present_in(files_list):
+            """Returns the index if the file is present"""
```
First line of doc string should have a period: ./favorites.py:298:1: D400 First line should end with a period.

facelessuser commented on favorite_files.py:
```diff
@@ -112,13 +112,17 @@ def add(self, names, group_name=None):
                     disk_omit_count += 1
+            else:
+                message = "is already among your favorite files and has not been added."
+                print(n, message)   # full filename printed to the console
```
Please use `error` which will use notifications or the Sublime error dialog for this.

If we want to also log something in the console, use something like this which will make it clear where the log came from:

```py
def log(msg):
    """Log message."""

    print("FavoriteFiles: %s" % msg)
```

facelessuser commented on favorite_files.py:
```diff
@@ -112,13 +112,17 @@ def add(self, names, group_name=None):
+            else:
+                message = "is already among your favorite files and has not been added."
+                print(n, message)   # full filename printed to the console
+                sublime.active_window().status_message("Some file " + message)
```
Why `Some file`?  Did you mean to use the base filename here?  Also, as mentioned above, please use `error`.

mg979 replied:
I forgot to explain this.. I wanted to avoid using error() because it triggers an error toolbox. More files can be already present and I wanted to avoid toolbox spamming. This only notifies in the status bar that some files haven't been added, and prints the details for the individual files in the console. The log() function would be better, I agree.

Anyway I added this primarily to test that everything was ok, I'm not meaning it should stay in the final version. If you find it appropriate it can stay, otherwise it can go. It's still some feedback on something has happened, I don't think it hurts, but using error() definitely would bother me, it can easily happen that there are duplicates if you add all files in the window. If you don't like it I remove it though, or I use log().

facelessuser replied:
Cool, that's fine for now.

mg979 committed:
fixed docstring, removed logging message for already present favorites

facelessuser commented:
Everything looks good so far 🙂 .

mg979 commented:
Thanks, sorry if I added unnecessary stuff, I'll keep it lean and clean from now on.

You can tell me how you'd like to proceed. Yesterday I did some work on the AddFavoriteFileCommand. Not that there's something broken about it, but it could be a bit better. It doesn't check for scratch files, and it asks to add current view even if it's not a valid file. I made it a bit more ocnsequential, but it's ot of critical importance. Here is a comparison with the functions I changed if you want to take a look:

https://gist.github.com/mg979/69727022ce3fb410c32565e00f75acc8

Other possible things I'd do are:

- remember per-project state
- setting to define a group from global favorites that should be added by default to all project favorites

Just ideas.

facelessuser commented:
> Yesterday I did some work on the AddFavoriteFileCommand. 

I'm okay with cleanup stuff. We'd have to do a formal review to see what gets in.  It's difficult for me to manually diff the gist and give you a line by line review of it.  A pull request is going to be the best way.

> remember per-project state

A pull request would be great as this is a currently open issue.

facelessuser commented:
As for this issue, we still need a way to change aliases.  I know that you plan to have the file manager to do this, but a simple command palette command is going to be needed to get merged.  As almost all the basic features are supported in some way via the command palette, it will be weird to not have this supported in some fashion.

I image a simple command to pick an entry and then get prompted with an input panel with the current alias.

That's it.  It should be nice and simple.  Once that is in, we can merge this, and then look at file manager etc.

mg979 commented:
About the alias command, for now I'd do it like this:

- when adding files, for each file evoke an input panel, pre-filled with the basename, so you can just hit enter if you don't want to give an alias. What you enter will be the current alias. This can have a setting, so that if can disable this prompt

- then maybe as you say a command from the palette to edit single aliases.

You prefer the second option only?

facelessuser commented:
I'm not against the first.  But we need at *least* the second.

mg979 commented:
Ok, I'll do the second first, then the first one if it's the case. Tomorrow, because now it's gotten late here.

facelessuser commented:
Yeah, no worries.  You don't need to rush it.  It'll get done when it gets done 🙂 .

mg979 committed:
Edit alias command from palette, optionally also when adding a single file(controlled from settings), remember toggle state across sessions

mg979 committed:
fixed for Travis

mg979 commented on favorite_files.py:
```diff
@@ -8,7 +8,7 @@
 import sublime_plugin
 from os.path import join, exists
-from FavoriteFiles.favorites import Favorites
+from FavoriteFiles.favorites import Favorites, FavFileMgr
```
I had to import this to set a key for the 'remember toggle state after restart'

mg979 commented on favorite_files.py:
```diff
@@ -50,6 +52,16 @@ def open_file(self, value, group=False):
+                    return
+                elif self.set_alias:
+                    return
+
```
I'm using the same SelectFavorite command to change the alias of a file

mg979 commented on favorite_files.py:
```diff
@@ -71,28 +83,34 @@ def fn(focus_view):
                 value -= self.num_files
-                self.files = Favs.all_files(group_name=self.groups[value][0].replace("Group: ", "", 1))
+                self.group_name = self.groups[value][0].replace("Group: ", "", 1)
+                self.files = Favs.all_files(group_name=self.group_name)
```
I'm passing group_name as an object variable so that it can be read in the new block, otherwise I had to repeat that expression in Favs.all_files()

mg979 commented on favorite_files.py:
```diff
@@ -71,28 +83,34 @@ def fn(focus_view):
                     self.window.show_quick_panel(
-                        [["Open Group", ""]] + self.files,
+                        [[string, ""]] + self.files,
                         lambda x: self.open_file(x, group=True)
```
Alternative string when changing aliases, only one file at a time is allowed.

mg979 commented on favorite_files.py:
```diff
@@ -122,6 +140,11 @@ def add(self, names, group_name=None):
+                # pass index as -1 because json will be sorted and we can't know it
+                # it will be searched in the called function
+                Favs.prompt_for_alias(-1, names[0], group_name)
+
```
it will ask for an alias when adding a file, if the setting is true

mg979 commented on favorite_files.py:
```diff
@@ -365,24 +388,32 @@ def run(self):
+                Favs.obj.files['project_mode'] = True
+                Favs.save(True)
             Favs.open(win_id=self.window.id())
 
```
This is used to remember toggle project state.

mg979 commented on favorite_files.py:
```diff
@@ -394,9 +425,18 @@ def check_st_version():
-    global Favs
-    Favs = Favorites(join(sublime.packages_path(), 'User', 'favorite_files_list.json'))
+    global_favs_file = join(sublime.packages_path(), 'User', 'favorite_files_list.json')
+    Favs = Favorites(global_favs_file)
```
I made a global for the global json file because it is used also in Toggle command

mg979 commented on favorites.py:
```diff
@@ -103,12 +103,24 @@ class FavFileMgr(object):
+        elif write:
+            with open(filename, "w") as file:
+                json.dump(data, file, sort_keys=True, indent=4, separators=(',', ': '))
+
```
I made a function for basic access to json file, that returns the contents or saves it. There were more places where the file had to be read and updated (check_aliases, and remember toggle state).

mg979 commented on favorites.py:
```diff
@@ -103,12 +103,24 @@ class FavFileMgr(object):
-            content = sanitize_json(f.read(), True)
-        data = json.loads(content)
+        data = cls.fav_file(favs, read=True)
 
```
Calls the function instead

mg979 commented on favorites.py:
```diff
@@ -118,8 +130,12 @@ def check_aliases(cls, favs):
+                data['project_mode'] = False
+
+            cls.fav_file(favs, data=data, write=True)
 
```
same here, it also adds the 'project_mode' key if not present, where 'toggle for project' state is stored.

facelessuser commented on favorites.py:
```diff
@@ -103,12 +103,24 @@ class FavFileMgr(object):
+                json.dump(data, file, sort_keys=True, indent=4, separators=(',', ': '))
+
+    @classmethod
     def check_aliases(cls, favs):
```
I didn't realize at the time, but this function should not be alias specific.  It should be called something like `upgrade_format` ans we may use it to upgrade to other formats in the future.

mg979 replied:
Ok upgrade_version?

mg979 commented on favorites.py:
```diff
@@ -165,10 +179,7 @@ def load_favorites(cls, obj, clean=False):
-                content = sanitize_json(f.read(), True)
-            file_list = json.loads(content)
+            file_list = cls.fav_file(obj.file_name, read=True)
 
```
more calls to fav_file() function

mg979 commented on favorites.py:
```diff
@@ -277,6 +288,48 @@ def add_group(self, s):
+            '''Find the index of the file that has been just added.'''
+            for f in node:
+                if f['file'] == name:
+                    return index, (f['alias'], name)
```
When a file is added, it alias defaults to basename, but it's still there, therefore I'm returning it too, it's the default alias that will be edited

facelessuser commented on favorites.py:
```diff
@@ -103,12 +103,24 @@ class FavFileMgr(object):
     """Handle file actions."""
 
     @classmethod
+    def fav_file(cls, filename, data=False, read=False, write=False):
```
Moving these to their own function is fine, but I'm not big on this function behaving different based on read or write.  In one case, it returns the data, in the other case it doesn't.  There isn't anything gained by having one function.  One called `read_fav_file` and one `write_fav_file` is fine.

mg979 replied:
Will split them

facelessuser commented on favorite_files.py:
```diff
@@ -394,9 +425,18 @@ def check_st_version():
+
 def plugin_loaded():
     """Setup plugin."""
+    global Favs, global_favs_file
```
I don't see a need for `global_favs_file`.  Can't this just be acquired from the Favs object?  Why a new dedicated global?

facelessuser commented on favorite_files.py:
```diff
@@ -394,9 +425,18 @@ def check_st_version():
+    Favs = Favorites(global_favs_file)
     check_st_version()
+
+    # restore toggle for project state if active
```
I think restoring projects should be in a separate pull as stated earlier.

mg979 replied:
Ok I'll remove the restore project state for now.

facelessuser commented on favorite_files.py:
```diff
@@ -8,7 +8,7 @@
 import sublime_plugin
 from os.path import join, exists
-from FavoriteFiles.favorites import Favorites
+from FavoriteFiles.favorites import Favorites, FavFileMgr
```
FavFileMgr was not originally intended to be exposed. It was supposed to be a helper for Favorites to perform actions.  Is it necessary to expose it here?  Can't we just add access functions to Favorites that relies on FavFileMgr?

mg979 replied:
I don't know if it's necessary, but I did it because it kept saving the key in the project json instead of the global one, when switching from project to global. If you don't want the import, I can just make file operations there, or put those read_file and write_file outside of FavFileMgr, it's not needed that hey stay inside there.

facelessuser replied:
I don't understand, what key is not going to the project specific settings?  And if this is related to remembering project specific settings, this shouldn't be in this pull as that is a separate feature that should be implemented in a separate pull.  There's is enough complexity going on in this pull as is.

mg979 replied:
I'll just remove this stuff for now

facelessuser commented on favorites.py:
```diff
@@ -277,6 +288,48 @@ def add_group(self, s):
+
+        w = sublime.active_window()
+        self.changing_alias_for = index, name, group_name
+        w.show_input_panel("Alias for this file:", name[0],
```
The sublime interface was meant to be handled in `favorite_files` not down  in this layer.

mg979 replied:
So I should move these functions in the other file? the problem is that they can be accessed by two different commands (SelectFavorite and AddFavorite)

mg979 replied:
I'll be away for a couple of hours, I'll push the changes later, to be sure I didn't break anything(removed this toggle thing, split the read/write file function). I just wanted to say that about the main thing, that is functions related to alias renaming, right now I can't see a different way than the one here, maybe I can make a new class in favorite_files.py, so both commands can use it, I'll work on it later.

facelessuser commented on favorite_files.py:
```diff
@@ -365,24 +388,32 @@ def run(self):
         if Favs.toggle_per_projects(win_id):
             error('Could not find a project file!')
         else:
+            # remember toggle for project state
```
Again, don't want to mix the feature for restoring project mode in this pull.

facelessuser commented on favorite_files.py:
```diff
@@ -365,24 +388,32 @@ def run(self):
 
         # Try and toggle back to global first
-        if not Favs.toggle_global(win_id):
+        if not restarted and not Favs.toggle_global(win_id):
```
This looks like its related to project toggling which is out of scope for this pull.  Also, FavFileMgr was not meant to be called directly from this layer.

facelessuser commented on favorite_files.py:
```diff
@@ -122,6 +140,11 @@ def add(self, names, group_name=None):
+            if len(names) == 1 and settings().get('always_ask_alias', False):
+                # pass index as -1 because json will be sorted and we can't know it
+                # it will be searched in the called function
+                Favs.prompt_for_alias(-1, names[0], group_name)
```
UI stuff should happen at this layer, and but acquiring aliases can be occur through `Favs`.

facelessuser reviewed (changes requested):
Here's my first pass.

mg979 committed:
removed toggle for project persistance, split function to read/write json, moved the prompt input function to the UI commands where it is called

mg979 committed:
no comment on this one

facelessuser commented:
I'll hopefully get back to reviewing this tomorrow.  Been a bit busy.

mg979 commented:
No problem, I didn't mean to stress you.

mg979 committed:
fix: check_plugin_version was misplaced causing an error

facelessuser commented on Default.sublime-commands:
```diff
@@ -12,6 +12,11 @@
         "command": "remove_favorite_file"
     },
     {
+        "caption": "Favorite Files: Edit alias for file",
```
Casing convention for commands should be followed.  Use title case.

facelessuser commented on favorite_files.py:
```diff
@@ -28,8 +28,23 @@ def run(self):
+
     def open_file(self, value, group=False):
-        """Open the file(s)."""
+        """Open the file(s). Also called when just editing aliases."""
```
Not sure we should be using the open file to edit alias.  I'll try and pull this down locally and take a better look at the approach.  I'm sure some of this could be cleaned up, but I want to play with it before I make any suggestions.

facelessuser commented:
I'll try and look at this tomorrow.  Most of it looks fine, but I'm going to take a closer look at some of the edit stuff.  I want see if I can come up with a little cleaner way to handle it.  I'll let you know once I've fully evaluated this.

mg979 commented on favorites.py:
```diff
@@ -218,6 +217,7 @@ def load_favorite_files(cls, obj, force=False, clean=False, win_id=None):
             else:
                 errors = True
 
+        cls.check_plugin_version(obj.file_name)
```
Forgot here to check for errors:

        # Update json if necessary
        if not errors:
            cls.check_plugin_version(obj.file_name)

mg979 commented:
If you check it locally, remember that it will change the json files, so maybe you'll want to make a backup of the file you'll be using.

facelessuser commented:
I've merged your changes and made some of my own here: https://github.com/facelessuser/FavoriteFiles/pull/19.  It was easier to to make the changes I was thinking then trying to explain what I was looking for 🙂 .  I'll merge alias support soon and then you can apply new work on top of that.

facelessuser commented:
Pull has been merged in commit: https://github.com/facelessuser/FavoriteFiles/commit/61dd3cd33ec8872580a3e218ec68e45f7e74a5dd.

facelessuser commented:
Additional work can be done on top of the recent commit on master.

mg979 commented:
Thanks :). While waiting I developed a bit some concepts, here is a demo of what I've done so far, you can tell me if you'd like that I keep working in this direction or if it's too much. More or less It's stable in my build, in the sense hat I won't be adding anything new for now(I have an idea, but for later).

![Imgur](https://i.imgur.com/YUJlCat.gif)

If it's too much or you aren't interested in this approach, I'll keep it as an external compatible project. Still it would need some changes in the main scripts. Namely, support for groups (when the file is saved, the group is remembered and when opened, if the group exists the file is opened in it). It would not require too many changes, but a few.

The Favorite Projects that you see relies on Project Manager for now(it was easier to do), it could be adapted to support Sublime project management only, too but I should work on it.

facelessuser commented:
If you have a concept I can try out, that may help.  I *think* I'm interested, but I'd like to try it to get an idea.  If it makes managing features etc. nicer, I'm definitely open.

mg979 commented:
Ok, I have to import all changes from master in my files, then I'll make a branch that you can test.

facelessuser commented:
It shouldn't be too bad.  It was mainly separating the edit command to a separate command, some bug fixes, and a little cleanup.  I did rename commands though to generally follow the convention I use in other plugins.  Everything else was document changes.

mg979 commented:
Unfortunately I also reformatted scripts for pep8, I'll take master as base and import changes instead.

facelessuser commented:
Everything in favorites should already be pep8 as well.

facelessuser commented:
Oh, you probably mean merge conflicts...

mg979 commented:
Just long lines. I can't handle merge conflicts, I still have to understand how to solve those. I always do hard reset when I get them.

facelessuser commented:
Yeah, I allow 120 lines.  I hate 80 char limit as it comes from the time people programmed on punch cards.  We don't program on punch cards anymore 🙂 .  Maybe we had this conversation already?  Can't remember, I complain every time someone tries to convince me I need to do 80 chars.  But it doesn't make sense in the modern high definition, wide screen monitor days (at least to me). 120 is not too long, but not too short which is just right 😉 .

mg979 commented:
I am on 1600*900 so I usually use 80 because I don't have to scroll horizontally, but mostly I force myself to respect that limit because the linter always complains. But it's not an issue, I'm ok with longer lines.

mg979 commented:
I finished editing and checking those files. Starting from current master, I added the new scripts and adapted the new changes in the old scripts(hoping I didn't forget anything). If you test it, it will change the json files again, but will make a backup (.json.version2).

https://github.com/mg979/FavoriteFiles/tree/dev_from_master

facelessuser commented:
Cool. I'll give it a try.

mg979 commented:
Sorry, I messed up something, must check it again.

mg979 commented:
It was ok, I messed up my mind only.

facelessuser commented:
I'm not against it, but its going to need a bit more polish.  Also, I wonder if it would be better if it wasn't directly editable, and instead was editable through indirect shortcuts.  The ability to edit it directly is quite buggy.  It would be easier to ensure format doesn't break if you do use indirect commands.  Like for instance, place cursor on alias, press alias edit shortcut, and use the input panel instead of editing directly.  You could use the same shortcut to move the files, but you'd let the command do it instead of relying on the general move line command.  I also don't see why we *need* to track a group number with the entries.

- One thing though, don't use `eval`.  Only use `eval` if there is no other way to do what you need, but in general, there is almost always another, and probably better way.

This has brought some things to my mind though.

- Aliases should probably have a predetermined format.  Probably `[\w\d-]`.  Maybe a little more, but we'll see.  But if anything is allowed, your format is going to get destroyed.

- Also, when editing aliases, I think if the user submits an empty value, the alias should be removed and the base file name should be used.

I think the above two points I will address before the next release.  I will probably go ahead and make a release with aliases as I suspect what you are working on will take a bit more time to mature, but I think it's a neat idea, and I am open to something like it, but there are a number of details that may need to be looked at closer.

mg979 commented:
> Also, I wonder if it would be better if it wasn't directly editable, and instead was editable through indirect shortcuts

I was considering this option, in fact the Favorite Projects board works like this. It's probably better.

> You could use the same shortcut to move the files, but you'd let the command do it instead of relying on the general move line command

It's possible to allow the swap_line command also in read only mode, it's how I'm using it in the projects board. It can't mess up anything that way.

> I also don't see why we need to track a group number with the entries.

It's not needed, but: usually you have certain kinds of files in a pane for a reason. If you add them and then reload the whole group, it makes sense to me that they go to the place they belonged to. Otherwise you must rearrange them manually, if they go in the first group. Anyway it can be eliminated, if you don't like it. Also note that the group is only restored if you open the whole group, single files are always added in the pane where you call the command.

> One thing though, don't use eval.

I read that eval is dangerous for user input, but are there other problems with it? I'm using it a lot, lately.

> Also, when editing aliases, I think if the user submits an empty value, the alias should be removed and the base file name should be used.

Yes absolutely.

> But if anything is allowed, your format is going to get destroyed.
I don't see the problem here. I couldn't destroy the format not even leaving them empty, right now they just show up empty. If you put empty spaces, those are remembered. I mean, I can't see what problems it may cause.

facelessuser commented:
> I don't see the problem here. I couldn't destroy the format not even leaving them empty, right now they just show up empty. If you put empty spaces, those are remembered. I mean, I can't see what problems it may cause.

If I add a name with `|`, it will mess up your format.  I think alias names should be restricted to something sane.

facelessuser commented:
And I mean restrict it from outside your project panel.  If I create a name outside your project panel, and you try and load it, and it is incompatible with your format, it will break things. Better to define rules for what a valid name is.

mg979 commented:
> If I add a name with |, it will mess up your format. I think alias names should be restricted to something sane.

Understood, I didn't try that. That could be solved in one way or another, but I'll make a read-only version anyway.

> And I mean restrict it from outside your project panel.

The restriction I'd put is a length restriction, and some character filter before it is saved in the file (the | is an obvious example). But I wouldn't be too strict for the rest.

facelessuser commented:
The character filter is what I'm talking about, and yes that would be applied before being saved.  I don't really care about length, but we could restrict that as well.

It is also much easier to be too restrictive and loosen it over time than not strict enough and tighten it and break a bunch of people's configurations.  I'd prefer to go a bit stricter than necessary, and see what people complain about not being able to use.  If nothing, I may try to restrict it to something like common file file name restrictions.

mg979 commented:
> The character filter is what I'm talking about

I'll just like to keep it simple. Then for me, only a-z0-9, all special characters replaced with whitespace, and max length 40 or so, both when added from the quick panel or from the view.

facelessuser commented:
For me, I'd prefer not to exclude foreign word characters as there is no reason to.  I'd probably open it up to Unicode word characters opposed to just ASCII word chars with a simple regex of `[\w\d\-]+` (at least in python, when enabling Unicode in the regex pattern you can do this).

If you are doing this in a syntax file, I believe you can use the posix character group (which I believe accounts for Unicode in syntax files) `[[:alpha:]\d_\-]+` to get Unicode alpha chars numbers and underscore. Again, I'm meh on length.  40 may be fine to start.  It could be increased if needed later.

facelessuser commented:
Obviously we could add spaces too.

facelessuser commented:
Not a big fan of replacing chars.  I'd like to give them "what you type is what you get".  But I would strip prefixed and trailing whitespace.

mg979 commented:
Ok, I'll try to make a version that:

- applies the filter for `[\w\d\ -]+` on save (stripping prefixed and trailing whitespace).
- restores basename if for some reason the alias is empty.
- doesn't use groups so probably I don't even need to to touch the main scripts
- defaults to read-only mode, with commands to edit aliases
- drops projects board for now(it needs Project Manager and it can't be a requirement).

Ok?

facelessuser commented:
And drop `eval` use.  99% of the time there is a better than using `eval`.  Yes it is evil in inputs, but in other cases, there are clearer ways to do the same thing 99% of the time.

I'm not sure how I feel about custom color schemes for the view either, I use dark themes, and find it jarring to have a tab suddenly decide to use a light scheme.  I'd prefer if it just used the default scheme.  If you are really attached to having a specific scheme attached to it, I'd make it configurable, and have it use the default scheme by default.

Anyways, I applied alias restrictions on the main branch: https://github.com/facelessuser/FavoriteFiles/commit/445090e78de2e4a4e6e4b4f186391c5ed3c226c7.  Will probably test it out a bit more and release it in the next couple of days.

mg979 commented:
I'll leave the restrictions to you, then I'll just use the same ones. I'll drop eval, I read a bit more about it so I won't use it.

About the color scheme, a custom scheme is needed because the syntax must be recognized, plain text would be ugly, and in the sidebar you need the line highlighted, I wouldn't like to be told "paste this in your scheme", but there are just a couple of patterns, I can make some variations with different colors or based on popular schemes, then it can be selected in the settings. If there is an easy way to do it, after there is a set to choose from, there can be an "auto" mode that defaults to dark/light depending on the scheme you're using.

Now I'm leaving for some hours, I'll work on it later or more probably tomorrow.

facelessuser commented:
Why can't you use common scopes?  `string`, `comment`, etc?  I guess I'm not sure why there has to be a specific scheme.  For [HexViewer](https://github.com/facelessuser/HexViewer) I use common scopes to style its special views.  Yeah, I have a couple extras scopes you could add to your color scheme, but the basics use common scopes.  Anyways, this can be explored in development.

mg979 commented:
Mainly, groups and files should look very differently, I'm using a different background for groups, I can't do that with common scopes. But I can use common scope names in the syntax and highlight them differently in the theme, and if the user chooses another scheme it won't look flat anyway.

facelessuser commented:
It's nothing we need to worry about now.  If we have to use a custom scheme, then that is fine.  I just want to explore it before we settle on shipping custom schemes.

facelessuser commented:
I'm going to back off character and size restriction as the real file name may not be restricted and still break the format.  So I won't worry about it.

mg979 commented:
Anyway aliases can be anything, even empty, and the file would be loaded just the same. They can't cause problems with the quick panel, at most with the special view, but there are ways to address that situation. For example, if  they are entered with an input panel, what is displayed must not represent the whole content. In that case, not even character `|` would break the format because the format wouldn't matter, the path wouldn't be read from the view anymore, and not even the alias itself. I don't think it's a real problem for now. The important thing is that it doesn't break the path, it shouldn't be hard to ensure that, even for an editable view.

Also, I don't know if you only tried the full view or also the sidebar(default key is f1), there you can't edit anything, it's read only, only `enter` and double click work.

facelessuser commented:
I wasn't worried about aliases breaking the command palette.  I was more worried about them breaking your new view, but then I realized you show the alias or the original base file.  I think Linux and macOS only restrict `:` from their paths (on the current file system), while Windows restricts a bit more.  But what had occurred to me was even if I restrict aliases, if no alias is applied, the base file name could break the format, so I decided not to worry about it.

I tired the sidebar briefly.  I switched into per project, and got stuck unable to switch back, then I switched into sidebar and got stuck in sidebar mode unable to switch back to the other mode.  I got frustrated and stopped testing as I noted much work still needed to be performed.  I was addressing aliases has a last step before I release that part, but now I realize restricting aliases won't really help what you are doing, but I wanted to at least add back the ability to restore the base file name when an empty aliases is submitted.

Anyways, as far as I am concerned, I think the alias feature is complete.  I'll let you keep working on this manager, and we'll see how it matures and see if we get can get it in or not.

mg979 commented:
> I tired the sidebar briefly. I switched into per project, and got stuck unable to switch back, then I switched into sidebar and got stuck in sidebar mode unable to switch back to the other mode

I'm sorry, it requires Project Manager plugin, it happens that way if you don't have it, I forgot to say.

> Anyways, as far as I am concerned, I think the alias feature is complete. I'll let you keep working on this manager, and we'll see how it matures and see if we get can get it in or not.

Yes it's better. I'll make a simpler version that doesn't modify the other scripts, that works in read only mode and with more schemes, if I have it working without issues I'll report back.
